File: tests/basic/loop_fusion.ml
open Optitrust
open Target


let _ = Run.doc_script_cpp (fun _ ->
  
  !! Loop_basic.fusion_on_block [cLabel "tofuse"];
  
)

"
int main() {
  int x;
  int y;
  tofuse: {
    for (int i = 0; (i < 5); i++) {
      x += i;
    }
    for (int i = 0; (i < 5); i++) {
      y += i;
    }
  }
}
"

let _ = Run.script_cpp ( fun _ ->

  !! Loop_basic.fusion_on_block [cLabel "tofuse"];

)

File: tests/basic/loop_fusion.cpp
int* t;
int* u;
int* v;
int n;

int main() {
  float p = 5.0;
  for (int i = 0; i < n; i++) {
    t[i] = i;
  }
  for (int i = 0; i < n; i++) {
    u[i] += i;
  }
  for (int i = 0; i < n; i++) {
    v[i] += i;
  }
  return 0;
}

int fusion_on_block() {
  float p = 5.0;
  for (int i = 0; i < n; i++) {
    t[i] = i;
  }
  for (int i = 0; i < n; i++) {
    u[i] += i;
  }
  for (int i = 0; i < n; i++) {
    v[i] += i;
  }
  return 0;
}
